Adding Rotation Logic to Renderer for use with Interpolated Facings

AspectInteractive2 opened the pull request:
This PR adds the ability for sprites to be rotated by the renderer, in order for additional interpolated facings to be created in between existing facings.

E.g.: 16 facings can be interpolated to 32 by adding an additional 16 interpolated facings. These interpolated facings are derived from the existing faces, by applying a fixed rotation that is proportionate to the gap between each facing.

***Add Interpolated Facings***
YAML flag: InterpolatedFacings
YAML options: Any number between 2 and 1024, that is greater than the number of Facings, and a power of 2.

Since sprites typically only have a set number of facings (e.g. 16 or 32), units that can rotate or move in more directions than they have facings will not have smooth movement between these facings. However, if we are able to rotate the sprite graphic between facings, we can create 'intermediate' facings between facings, as shown below. The code here takes the delta between the actual facing and the next intended interpolated facing (depending on how many are required), and applies this as a rotation to the sprite. The most common number of intended interpolated facings is 1024, as this is the maximum number of facings, however any number that is greater than the existing number of facings, that is also a power of 2 can be used as the InterpolatedFacings amount.

This rotation logic was used in the original red alert to smoothen the rotation of all aircraft with 16 facings. So to mimic the original game, I have added this flag to all aircraft that currently have 16 facings. As you can see in the below video, the movement is much smoother once this flag is enabled (right side of the video). By request I have also added this to the carry-all in D2K, in an attempt to smoothen jitters this unit appears to have in the current build.

https://www.dropbox.com/s/u7cucyvd3hjvti0/mig_orig_interp_compare.mp4?dl=0

A second use for this same feature is to use just 1 facing, and rely entirely on the InterpolatedFacings flag to create all remaining faces (e.g. 16, 32, or any number up to 1024). This enables sprites to rely entirely on the renderer's new rotation for their facings. While this is not as suitable for units that have perspective (since the shadows and lighting are not rotated, it will look a bit off), it is highly useful for flat or top-down viewed objects such as projectiles, or top-down units such as in the indie RTS Rusted Warfare (in fact this feature was my original motivation for this PR). This solves the following issue: https://github.com/OpenRA/OpenRA/issues/19146. This also greatly reduces the time required to develop such sprites, since only 1 facing (north) needs to be produced instead of 16 or 32. The below video shows what this feature may look like applied to the existing medium tank. While the tank here is not fully suited to the feature, it shows that the tank can appear and function almost like the original, despite only 1 facing being used (since all other facings are rotations).

https://www.dropbox.com/s/qsfa0fv7oe3dtp0/2tnk_userotationsforfacings.mkv?dl=0

**Caveat**
All of my tests have been in Red Alert and D2K. While the feature is not needed for other mods, I am unsure how it would interact with them, e.g. if there would be any unintended side effects. I don't think this is likely, but it is still possible.

abcdefg30 reviewed (commented):
Three questions:
- Can we undo several of the line breaks?
- Why did you declare all `int` parameters either `in` or `out`? Is that really necessary?
- Can you please use tabs instead of spaces for the yaml indentation?

AspectInteractive2 commented:
> Three questions:
> 
> * Can we undo several of the line breaks?

So my rationale is that now that DrawSprite takes an additional parameter, it is not possible to fit the full method signature on one line without leading to scrolling. Now not all calls to DrawSprite use the additional parameter, but I thought it would be inconsistent to have some of the DrawSprite methods on two lines and others on one, so I opted to put all of them on two lines. Also this means if the layer or rotAngle parameters need to be added at a later time, no extra line breaks needed to be used. But if you need it, of course I can change it back for the DrawSprite calls that don't currently use the rotAngle parameter.

> * Why did you declare all `int` parameters either `in` or `out`? Is that really necessary?

The 'in' modifier was in part to match the existing parameters that DrawSprite uses, such as a,b,c,d,tint,size,alpha. But I think it is good practice, because this highlights that no downstream method can modify the upstream parameters, which is especially important for the renderer since the renderer should not alter logic upstream (I suspect this is why they were added).

The 'out' parameter has to be used in order for the DefaultSpriteSequence to be able to modify the Animation's rotation angle, while also returning the sprite. The alternative would be to create a struct with the Sprite and the rotAngle, but that would be messier and wouldn't provide much benefit apart from omitting the 'out'. Also this method signature was suggested by pchote initially, and I ended up finding it appropriate for the task.

> * Can you please use tabs instead of spaces for the yaml indentation?

I thought I did but happy to change this.
EDIT: This has now been fixed.

pchote commented on OpenRA.Game/Graphics/RgbaSpriteRenderer.cs:
```diff
@@ -22,28 +22,28 @@ public RgbaSpriteRenderer(SpriteRenderer parent)
 		}
 
-		public void DrawSprite(Sprite s, in float3 location, in float3 scale)
+		public void DrawSprite(Sprite s, in float3 location, in float3 scale, in int rotAngle = 0)
```
Please change these to `public void DrawSprite(Sprite s, in float3 location, in float3 scale, float rotation = 0)` where `rotation` is given in radians. The renderer works in physical units, not simulation ones, so we shouldn't be using WAngle here (and *especially* not WAngle stored in an int!).

AspectInteractive replied:
Done

reaperrr replied:
Seems this hasn't been changed yet...?

pchote commented on OpenRA.Game/Graphics/SequenceProvider.cs:
```diff
@@ -36,7 +36,7 @@ public interface ISpriteSequence
 
 		Sprite GetSprite(int frame);
-		Sprite GetSprite(int frame, WAngle facing);
+		Sprite GetSprite(int frame, WAngle facing, out int rotAngle);
```
I suggest returning WAngle here and having `SpriteRenderable` do the conversion to radians, like it already does for WPos to pixels.

AspectInteractive replied:
Done

pchote commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -17,30 +17,34 @@ namespace OpenRA.Graphics
 		readonly int zOffset;
 		readonly PaletteReference palette;
 		readonly float scale;
+		readonly int rotAngle;
```
`readonly WAngle rotation`.

AspectInteractive replied:
Done

pchote commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -53,16 +57,21 @@ public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, Palet
-		public IRenderable OffsetBy(in WVec vec) { return new SpriteRenderable(sprite, pos + vec, offset, zOffset, palette, scale, alpha, tint, tintModifiers, isDecoration); }
-		public IRenderable AsDecoration() { return new SpriteRenderable(sprite, pos, offset, zOffset, palette, scale, alpha, tint, tintModifiers, true); }
+		public IPalettedRenderable WithPalette(PaletteReference newPalette)
+		{ return new SpriteRenderable(sprite, pos, offset, zOffset, newPalette, scale, alpha, tint, tintModifiers, isDecoration, layer); }
```
Function definitions must either be all in a single line as these were previously defined, or the opening and closing braces must be on their own lines.

These must also all pass the rotation into the new constructors to avoid regressions in e.g. the target flash or iron curtain overlays.

AspectInteractive replied:
Have now split the function onto separate lines, and added the new rotation variable to all of the calls to prevent future regressions.

pchote commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -93,7 +102,7 @@ public void Render(WorldRenderer wr)
 				a *= -1;
 
-			wsr.DrawSprite(sprite, palette, ScreenPosition(wr), scale, t, a);
+			wsr.DrawSprite(sprite, palette, ScreenPosition(wr), scale, t, a, rotAngle);
```
WAngle to radians conversion here.

AspectInteractive replied:
This has now been done.

pchote commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -126,38 +126,43 @@ float ResolveTextureIndex(Sprite s, PaletteReference pal)
 			var samplers = SetRenderStateForSprite(s);
-			Util.FastCreateQuad(vertices, location + scale * s.Offset, s, samplers, paletteTextureIndex, nv, scale * s.Size, float3.Ones, 1f);
+			Util.FastCreateQuad(vertices, location + scale * s.Offset, s, samplers, paletteTextureIndex, nv,
+								scale * s.Size, float3.Ones, 1f, rotAngle);
```
Remove strangely placed linebreaks here and below.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -20,12 +20,26 @@ public static class Util
+			{
+				var centerPoint = o + (size / 2);
+				var rotWAngle = new WAngle(rotAngle);
+				a = RotatePoint(o, centerPoint, -rotWAngle);
```
`FastCreateQuad` is a hot path so we want to avoid unnecessary overhead.

This can be done much more efficiently by calculating the center point, calculating the rotated delta-x and delta-y once, and then adding the deltas in the different combinations to get each of the four corners.

AspectInteractive replied:
I have moved the Sin and Cos calculations out of `RotatePoint()` as they were being calculated multiple times when they only need to be calculated once. This leaves a few multiplications and additions within the `RotatePoint()` function, which I don't think can be brought out easily, as the deltas are not the same for each point. It may be possible with some significant rework, but my math is not quite at the level to figure this out. Still I think with the Sin and Cos being only calculated once, this version should be very close to optimal.

pchote commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -195,6 +197,8 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 				var flipY = LoadField(d, "FlipY", false);
 
 				Facings = LoadField(d, "Facings", 1);
+				InterpolateFacings = LoadField(d, "InterpolateIntermediateFacings", false);
```
Thinking about this more, I now disagree with my original advice for two reasons:
* Sprites distort in an ugly way when rotated by a small amount, and continuous rotation looks odd when combined with quantized rotation from other sprites.
* There are several parts of the code (e.g. `BodyOrientation`, `Turreted`) that rely on the sequence facing number for quantizing simulation-related rotations. These will be broken by the current implementation.

A better solution would be to define an `InterpolatedFacings: <integer>` field, have `GetSprite` quantize to the nearest interpolated facing, and introduce a new method that `BodyOrientation` etc can call to return the appropriate facing value depending on how the sequence is defined.

`UseRotationsForFacings` should then be redundant against defining `Facings: 1` with e.g. `InterpolatedFacings: 32`.

AspectInteractive replied:
Good idea, in the new version I now have the one `InterpolatedFacings` attribute, which works as you describe, by quantizing the original rotation from `AngleDiffToStep()` to the nearest interpolated facing. In this way `Facings: 16, InterpolatedFacing: 32` will display 32 facings, 16 of which come from the original `Facing` attribute, while the remaining 16 are created by applying rotation. Then to get the same behaviour as shown in my original PR, you simply set `InterpolateIntermediateFacings: 1024`, which I have done in the latest version. This works because 1024 facings makes the step size 1, which prevents the quantizing from reducing the number of rotations. Also this now can be used in place of `UseRotationsForFacings` just as you describe, by simply setting `Facings: 1` and `InterpolatedFacings: n` where n is however many interpolated facings you want. Have tested this and it seems to work, with one small caveat, that having a Facing of 1 does not seem to be compatible with the `UseClassicSprites` flag due to it requiring 32 facings. But I suspect that isn't important since the UseClassicSprites flag becomes unnecessary if only one facing is used.

AspectInteractive commented:
Thanks @pchote for the feedback, will look to make the above changes. I am fairly busy these days so not sure when I will have the time (have some personal matters to deal with as well), but will do it in due course.

reaperrr commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -17,30 +17,34 @@ namespace OpenRA.Graphics
 	public class SpriteRenderable : IPalettedRenderable, IModifyableRenderable, IFinalizedRenderable
 	{
 		public static readonly IEnumerable<IRenderable> None = new IRenderable[0];
-
```
I think that newline can stay (that would also have avoided the rebase conflict).

reaperrr commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -17,30 +17,34 @@ namespace OpenRA.Graphics
 
-		public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, PaletteReference palette, float scale, float alpha, float3 tint, TintModifiers tintModifiers, bool isDecoration)
+		public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, PaletteReference palette, float scale, float alpha,
+								float3 tint, TintModifiers tintModifiers, bool isDecoration, WAngle rotation, int layer = 0)
```
Our general practice is to indent places like this with only a single tab.

reaperrr commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -49,29 +53,49 @@ public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, Palet
 		}
 
+		public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, PaletteReference palette, float scale, float alpha,
+								float3 tint, TintModifiers tintModifiers, bool isDecoration, int layer = 0)
```
Same here.

reaperrr commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -126,38 +126,44 @@ float ResolveTextureIndex(Sprite s, PaletteReference pal)
 
-		internal void DrawSprite(Sprite s, float paletteTextureIndex, in float3 location, float scale, in float3 tint, float alpha)
+		internal void DrawSprite(Sprite s, float paletteTextureIndex, in float3 location, float scale, in float3 tint, float alpha,
+								 float rotation = 0F)
```
Same here.

reaperrr commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -126,38 +126,44 @@ float ResolveTextureIndex(Sprite s, PaletteReference pal)
 
-		public void DrawSprite(Sprite s, PaletteReference pal, in float3 location, float scale, in float3 tint, float alpha)
+		public void DrawSprite(Sprite s, PaletteReference pal, in float3 location, float scale, in float3 tint, float alpha,
+							   float rotation = 0F)
```
Same here.

reaperrr commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -126,38 +126,44 @@ float ResolveTextureIndex(Sprite s, PaletteReference pal)
 
-		internal void DrawSprite(Sprite s, float paletteTextureIndex, in float3 a, in float3 b, in float3 c, in float3 d, in float3 tint, float alpha)
+		internal void DrawSprite(Sprite s, float paletteTextureIndex, in float3 a, in float3 b, in float3 c, in float3 d, in float3 tint,
+								 float alpha)
```
Same here. Actually, this doesn't look like anything changed except moving float alpha to the next line? In that case this can probably just be reverted.

reaperrr commented on OpenRA.Game/Graphics/UISpriteRenderable.cs:
```diff
@@ -22,8 +22,10 @@ public class UISpriteRenderable : IRenderable, IPalettedRenderable, IFinalizedRe
 
-		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette, float scale = 1f, float alpha = 1f)
+		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette,
+								  float scale = 1f, float alpha = 1f, float rotation = 0F)
```
Same here.

reaperrr commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -20,12 +20,27 @@ public static class Util
 
-		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv, in float3 size, in float3 tint, float alpha)
+		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv,
+										  in float3 size, in float3 tint, float alpha, in float rotation = 0F)
```
And here.

reaperrr commented on mods/d2k/sequences/aircraft.yaml:
```diff
@@ -2,7 +2,7 @@ carryall:
 		Start: 1923
 		Facings: -32
-		InterpolateIntermediateFacings: True
+		InterpolatedFacings: 1024
```
Out of curiosity, will these interpolated facings be preloaded (added to the game's sprite sheets) or does the engine just rotate the 32 actual sprite frames on the fly, without any impact on the number of sheets/VRAM usage? I'm not good enough at understanding the code to tell from that.

AspectInteractive replied:
Yes it rotates the 32 actual sprite frames on the fly to obtain the remaining 992 facings - no additional sprite sheets, facings or otherwise need to be loaded.

reaperrr commented on mods/ra/sequences/aircraft.yaml:
```diff
@@ -1,6 +1,7 @@
+		InterpolatedFacings: 1024
 	icon: migicon
 
 yak:
```
Looks like you forgot the Yak?

And if facing interpolation doesn't impact sheets/RAM usage, I think we should also use 1024 for helicopters (in both RA and TD) as well as any other TD aircraft with 32 facings, because the jittering is present there too, just less pronounced than with RA's 16-facing aircraft.

AspectInteractive replied:
Yes that was an oversight, have now added it to Yak as well. The reason it wasn't added to the helicopters is because the original Red Alert only had this rotation effect for 16 facing aircraft, and helicopters have 32 facings. It can still be added of course, but may be better to get some consensus on this since it deviates from the original game a bit. I have made an exception for the Carry-All from D2K for this jittering reason, so in my view it makes sense to do it for the helicopters as well if it improves the look of the animation.

reaperrr commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -136,6 +136,7 @@ public class DefaultSpriteSequence : ISpriteSequence
 		public int Length { get; private set; }
 		public int Stride { get; private set; }
 		public int Facings { get; private set; }
+		public int InterpolatedFacings { get; private set; }
```
I think we'll want this to be implemented on `SequenceProvider.ISpriteSequence`, so that places like `QuantizeFacingsFromSequence` can look up the value.
`QuantizeFacingsFromSequence` will then want to use `InterpolatedFacings` rather than `Facings` if its value is higher than `Facings`, otherwise the orientation of offsets will look off/coarse versus the sprite when actors turn (set a higher `InterpolatedFacings` on the RA cruiser `idle` sequence to see what I mean, the turrets will 'jump' because their position offset relative to the body is still calculated with 16 facings unless `QuantizeFacingsFromSequence` is fixed to account for that).

AspectInteractive replied:
This is a great suggestion, I have now implemented it and it has indeed smoothened the cruiser turret animation quite a bit, as shown below. There appears to be some glitchiness with the ship that isn't present with the aircraft, which I am guessing is because the original sprites are not at perfect angles relative to their facing, so it is probably not worth using InterpolatedFacings on the ships (or perhaps the number of facings should be reduced). But this change does now mean that turret facings interpolate with the unit itself, which is quite nice.

https://www.dropbox.com/s/9tu426nuynz1ugc/quantize_interp_facings.mp4?dl=0

reaperrr commented on OpenRA.Game/Graphics/RgbaSpriteRenderer.cs:
```diff
@@ -22,28 +22,28 @@ public RgbaSpriteRenderer(SpriteRenderer parent)
 		}
 
-		public void DrawSprite(Sprite s, in float3 location, float scale = 1f)
+		public void DrawSprite(Sprite s, in float3 location, float scale = 1f, in int rotAngle = 0)
```
Same here...

reaperrr commented on OpenRA.Game/Graphics/RgbaSpriteRenderer.cs:
```diff
@@ -22,28 +22,28 @@ public RgbaSpriteRenderer(SpriteRenderer parent)
 		}
 
-		public void DrawSprite(Sprite s, in float3 location, float scale, in float3 tint, float alpha)
+		public void DrawSprite(Sprite s, in float3 location, float scale, in float3 tint, float alpha, in int rotAngle = 0)
```
...and here.

AspectInteractive commented:
I have addressed all the feedback received so far, as well as rebased to the latest version of ORA as of yesterday.

reaperrr commented:
Huh? Why was this closed?

AspectInteractive2 commented:
> Huh? Why was this closed?

It closed when I hard reset the branch, I had some issues with rebasing. Have re-opened it now and it is up to date.

reaperrr commented on OpenRA.Game/Graphics/Animation.cs:
```diff
@@ -50,18 +51,22 @@ public Animation(World world, string name, Func<WAngle> facingFunc, Func<bool> p
-			var imageRenderable = new SpriteRenderable(Image, pos, offset, CurrentSequence.ZOffset + zOffset, palette, CurrentSequence.Scale, alpha, float3.Ones, tintModifiers, IsDecoration);
+			var imageRenderable = new SpriteRenderable(Image, pos, offset, CurrentSequence.ZOffset + zOffset, palette,
+													   CurrentSequence.Scale, alpha, float3.Ones, tintModifiers, IsDecoration, rotation,
+													   0);
```
To be honest, I feel this isn't worth putting on an extra line just for readability/optics (assuming that was the reason).

AspectInteractive2 replied:
Actually the reason is it would be too long and require horizontal scrolling otherwise, the readability is purely an unintended side effect:

![image](https://user-images.githubusercontent.com/1495872/151649621-65f06658-8d2f-4ae4-87ab-47402e8ef81e.png)

reaperrr commented on mods/ra/sequences/ships.yaml:
```diff
@@ -6,6 +6,7 @@ ss:
 ca:
 	idle:
 		Facings: 16
+		InterpolatedFacings: 1024
```
This seems to have slipped in from testing, and I agree it doesn't look that good on ships due to the angles their facings were rendered at, so this can be removed.

AspectInteractive2 replied:
Sure, I have removed it, though perhaps it could be included with a smaller number of facings, e.g. 32 or 64. I will leave it to others to test on their own though, rather than having it in the PR.

AspectInteractive2 replied:
One more thing, I noticed that for quantized facings such as turrets, the interpolated facings don't rotate at the exact time as the normal facings, as you can see below. I suspect this might be because QuantizeFacingsFromSprites() runs either before or after InterpolatedFacings are calculated, making them 1 frame off. I'm not sure how to fix this however, since I suspect delaying or speeding up the calculation of InterpolatedFacings is not straightforward (and could have side effects on the Mig and other aircraft that don't use QuantizeFacigsFromSprites()).

https://www.dropbox.com/s/zuw4asuvt7kg3lw/staggered_facings.mkv?dl=0

reaperrr commented:
As far as I'm concerned, this looks good to me once the commits have been squashed (note: something still seems to be a bit wrong with how it was rebased, there seem to be some outdated duplicate commits in the commit history).

> One more thing, I noticed that for quantized facings such as turrets, the interpolated facings don't rotate at the exact time as the normal facings, as you can see below. I suspect this might be because QuantizeFacingsFromSprites() runs either before or after InterpolatedFacings are calculated, making them 1 frame off. I'm not sure how to fix this however, since I suspect delaying or speeding up the calculation of InterpolatedFacings is not straightforward (and could have side effects on the Mig and other aircraft that don't use QuantizeFacigsFromSprites()).

My idea would be to add a `Rotation` field for sequence frames listed under `Combine`, to give the modder more manual control over which frame to use as base and by how much (and in which direction) to rotate them for additional facings, and then list those along with the default facings. This could also work around the delay issue, as the additional frames would be baked into the sequence definition.
I mean like this:
```
ca:
	idle:
		Combine:
			ca:
				Frames: 0
			ca:
				Frames: 0
				Rotation: -32
			ca:
				Frames: 1
			ca:
				Frames: 1
				Rotation: -32
			(...)
			ca:
				Frames: 15
			ca:
				Frames: 0
				Rotation: 32
		Facings: 32
```

But I think that should be left for a follow-up PR, let's get this one merged first.

AspectInteractive2 commented:
Thanks, yes in a future PR we can expand this further to allow custom rotations for each individual facing.

I have rebased again with bleed and squashed all commits, so it is ready to go if there are no further changes needed.

reaperrr commented:
I could really use this downstream, so unless anyone finds some fundamental flaw with the implementation, I'd like to see this merged soon. Getting this into the playtest can't hurt either.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -25,23 +25,28 @@ public class SpriteRenderable : IPalettedRenderable, IModifyableRenderable, IFin
 		readonly TintModifiers tintModifiers;
 		readonly float alpha;
 		readonly bool isDecoration;
+		readonly int layer;
```
What is `layer`? It doesn't appear to be used and is always set to 0 AFAICT?

AspectInteractive2 replied:
Ah, this appears to have been a carry-over from my own fork of the engine, where I've added z-ordering of renderables via a 'layer' variable. I may look to add this in a separate PR, as it's useful for overlays in particular, but it has no relevance to the rotation, so I've now removed it.

RoosterDragon commented on OpenRA.Game/Graphics/Animation.cs:
```diff
@@ -50,18 +51,21 @@ public Animation(World world, string name, Func<WAngle> facingFunc, Func<bool> p
 		public int CurrentFrame => backwards ? CurrentSequence.Length - frame - 1 : frame;
-		public Sprite Image => CurrentSequence.GetSprite(CurrentFrame, facingFunc());
+
+		public Sprite Image => CurrentSequence.GetSprite(CurrentFrame, facingFunc(), out rotation);
```
The use of the `out` parameter here is making me nervous, as we're relying on side effects on calling this property. If we reorder the parameters to `SpriteRenderable` we could get the wrong result.

Plus, for most callers to `Image` determining the rotation is a calculation we then throw away.

Would it make more sense to keep the current signature of `GetSprite` and introduce a new `GetRotation` that we can call specifically in the `Render` method?

AspectInteractive2 replied:
The problem with introducing a separate call is that the rotation is very tightly coupled to the frame and facing parameters used in the GetSprite() call. Both of these parameters need to be used to calculate the rotation, meaning that moving the rotation out of GetSprite() would make the design much more complicated. Also from what I can see the Image function is not only used by the Render function, but also by other classes. Therefore moving the rotation to the Render() function would prevent future use of the rotation property by other classes that do not call Animation's Render() function. This current design was also suggested by pchote, and I have found it quite intuitive, so I would be reluctant to move away from it unless there is a strong reason to do so.

>Plus, for most callers to Image determining the rotation is a calculation we then throw away.

Actually the calculation only occurs for sprites with a rotation due to the 'if (InterpolatedFacings != -1)' check, so for 99% of sprites the rotation will never be calculated.

RoosterDragon replied:
The `rotation` is currently only used within the `Render` method, so the design that I am imagining is something roughly like:
```c#
public Sprite Image => GetImageAndRotation().Image;

(Sprite Image, WAngle Rotation) GetImageAndRotation()
{
    var frame = Frame;
    var facing = facingFunc();
    return (CurrentSequence.GetSprite(frame, facing), CurrentSequence.GetRotation(frame, facing));
}

public IRenderable[] Render(...)
{
    var (image, rotation) = GetImageAndRotation();
    var imageRenderable = new SpriteRenderable(image, ..., rotation);
    ...
}
```

If we ever did worry about the overhead of interpolating facings for those sprites that use them, we could add a parameter to skip it e.g.  `GetImageAndRotation(bool shouldCalculateRotation)`.

The concrete benefits are:
- Don't need to store an extra field in `Animation` for rotation.
- The `Render` method no longer relies on the side effect that `Image` must be called before `rotation`. So if anybody ever swaps those around in a seemingly safe refactoring they won't break everything.
- Avoids the overhead of calculating the interpolated rotation when calling `Image` if we don't need it. (If we have an `shouldCalculateRotation` option)

AspectInteractive2 replied:
I see what you mean. Ok I have implemented the above, removing the rotation field from Animation and returning a tuple from GetSpriteWithRotation() (I used the word Sprite to be consistent with existing terminology, and because that's the class name). I also kept the original GetSprite() call for the Image function. Since GetSprite() is no longer overloaded, it can remain as is to reduce overhead, as we are now using GetSpriteWithRotation() exclusively for the SpriteRenderable creation.

RoosterDragon commented on OpenRA.Game/Graphics/UISpriteRenderable.cs:
```diff
@@ -22,8 +22,10 @@ public class UISpriteRenderable : IRenderable, IPalettedRenderable, IFinalizedRe
+		readonly float rotation;
 
-		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette, float scale = 1f, float alpha = 1f)
+		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette,
```
`WithPalette` fails to maintain the new `rotation`.

AspectInteractive2 replied:
Fixed

RoosterDragon commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -20,12 +20,27 @@ public static class Util
+			{
+				var centerPoint = o + (size / 2);
+				var angleSin = (float)Math.Sin(-rotation);
+				var angleCos = (float)Math.Cos(-rotation);
```
`Math.SinCos`? Hopefully Mono has it.

abcdefg30 replied:
What is the status here? Does Mono not have `Math.SinCos`?

AspectInteractive2 replied:
I think it does because the Math class is showing compatibility with Xamarin.iOS 10.8 and Xamarin.Mac 3.0 on the official Microsoft doc [here](https://docs.microsoft.com/en-us/dotnet/api/system.math?view=net-5.0) 
![image](https://user-images.githubusercontent.com/1495872/153021267-c9caa792-b9b5-47a5-960b-96cf1d632242.png), and it would be odd if they didn't support the Sin and Cos functions, but I will I also get this tested and report back. I can confirm it works on Linux (Manjaro distro) without issue.

AspectInteractive2 replied:
Pun has tested it on his new Mac and it works fine, as per the below video, meaning it works on all 3 OSes (Windows, Linux, Mac). Given this and the above compatibility table, I think we should be safe here.

https://user-images.githubusercontent.com/1495872/153030642-9ff5ee0d-4072-4d9d-8dab-ef376ab01fb4.mov

RoosterDragon commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -20,12 +20,27 @@ public static class Util
 
-		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv, in float3 size, in float3 tint, float alpha)
+		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv,
+			in float3 size, in float3 tint, float alpha, in float rotation = 0F)
```
I would drop the `in` keyword when passing `float`. https://docs.microsoft.com/en-us/dotnet/csharp/write-safe-efficient-code#use-in-parameters-for-large-structs says "performance may degrade by using pass-by-reference for types smaller than [IntPtr.Size](https://docs.microsoft.com/en-us/dotnet/api/system.intptr.size#system-intptr-size)."

The `float3` parameters can keep it.

AspectInteractive2 replied:
Fixed

RoosterDragon commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -248,6 +263,16 @@ public static float[] MatrixMultiply(float[] lhs, float[] rhs)
 		}
 
+		// Using this method: https://stackoverflow.com/questions/2259476/rotating-a-point-about-another-point-2d
+		public static float3 RotatePoint(in float3 point, in float3 centerPoint, in float angleSin, in float angleCos)
```
Drop `in` against `float` parameters.

AspectInteractive2 replied:
Fixed

abcdefg30 commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -50,6 +53,10 @@ public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, Palet
 
+		public SpriteRenderable(Sprite sprite, WPos pos, WVec offset, int zOffset, PaletteReference palette, float scale, float alpha,
+			float3 tint, TintModifiers tintModifiers, bool isDecoration)
+		: this(sprite, pos, offset, zOffset, palette, scale, alpha, tint, tintModifiers, isDecoration, WAngle.Zero) { }
```
```suggestion
			: this(sprite, pos, offset, zOffset, palette, scale, alpha, tint, tintModifiers, isDecoration, WAngle.Zero) { }
```

AspectInteractive2 replied:
Fixed

abcdefg30 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -195,6 +198,11 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 
 				Facings = LoadField(d, "Facings", 1);
+				InterpolatedFacings = LoadField(d, "InterpolatedFacings", -1);
+				if (InterpolatedFacings != -1 && !(InterpolatedFacings > 1 && InterpolatedFacings > Facings && InterpolatedFacings <= 1024
+					&& Exts.IsPowerOf2(InterpolatedFacings)))
```
```suggestion
				if (InterpolatedFacings != -1 && (InterpolatedFacings <= 1 || InterpolatedFacings <= Facings || InterpolatedFacings > 1024 || !Exts.IsPowerOf2(InterpolatedFacings)))
```
Since only one condition is needed to fail, using ORs makes this easier to read Imo. (And fits the wording you use in the exception message.)

AspectInteractive2 replied:
Makes sense. The only point I would say is given the fairly narrow set of valid numbers (9), describing the statements that are True seems more intuitive to me at least, but happy to go with the change as it can easily be argued both ways. Have updated it to your version.

abcdefg30 commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -67,11 +67,26 @@ public static int GetTurnDirection(WAngle facing, WAngle desiredFacing)
+		{
+			var step = 1024 / numFrames;
+			var a = (facing.Angle + step / 2) & 1023;
+			return new WAngle(a % step - step / 2);
```
Does it make sense to wrap this in a `new WAngle` considering the only usage immediately calls `.Angle`?

abcdefg30 replied:
Ah, I see. Turning this into a WAngle solves the problem of the value becoming negative. Not sure if that is a good solution.

AspectInteractive2 replied:
I think it also reads better to return an angle as a WAngle rather than a float or int. Previously I had passed ints instead of WAngles, but pchote recommended to pass WAngle instead, which I think is cleaner since it ensures the Angle is valid (in this case that negatives are solved for and the value is clamped to the range 0-1024).

abcdefg30 commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -67,11 +67,26 @@ public static int GetTurnDirection(WAngle facing, WAngle desiredFacing)
 			return a / step;
 		}
 
+		// Returns the remainder angle after rounding to the nearest whole step / facing
```
Missing `<summary></summary>`?

AspectInteractive2 replied:
Fixed I think? Have never used `<summary></summary>` before so let me know if it is incorrect.

abcdefg30 commented on OpenRA.Test/OpenRA.Game/GraphicUtilTest.cs:
```diff
@@ -0,0 +1,67 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
```
2022 :P

AspectInteractive2 replied:
Fixed

abcdefg30 commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -67,11 +67,26 @@ public static int GetTurnDirection(WAngle facing, WAngle desiredFacing)
 		/// </summary>
 		public static int IndexFacing(WAngle facing, int numFrames)
 		{
+			// 1024 here is the max angle, so we divide the max angle by the total number of facings (numFrames)
```
Tbh, I'm not sure what that comment is supposed to tell me which isn't covered by reading `1024 / numFrames`. So I'd remove it.

AspectInteractive2 replied:
I did not understand this function for some time, because 1024 is also used as the length of a cell (my first intuition), and in general is a fairly generic number being 2^10. In a perfect world all [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) would be commented, but I made a point to comment this one because despite having spent many months with this engine, this took me some time to figure out. You can remove it, but I think that is a step in the wrong direction, particularly as the codebase is already in my view too light on commentary, particular for the trickier parts such as this IndexFacing() function.

abcdefg30 replied:
Having this only here is really odd (especially considering the functions below do the same but don't have it). And putting the comment everywhere is just littering. A better solution would probably be introducing some sort of const `WAngle.MaxAngle` that is used instead of 1024 in a follow-up PR.

AspectInteractive2 replied:
That's probably the best solution, but in the absence of a constant, I think the comment is worth retaining. And I think the fact that IndexFacing() is the first facing related function in Util.cs, means anyone editing the class is likely to read the comment first even if they aren't modifying that specific function, so it doesn't need to be repeated in all subsequent functions. I am happy to look to replace this with WAngle.MaxAngle or similar in a future PR, if we can leave the comment here for reference until the constant is added. Also the point about numFrames being equal to the total number of facings is worth noting, as this wasn't immediately apparent either (since the term 'frames' is used for other purposes in other areas of the code).

AspectInteractive2 commented:
Hello, just wondering if there is an update on this, as I would like to get this merged, in part so I don't have to keep rebasing.

abcdefg30 commented:
https://github.com/OpenRA/OpenRA/pull/19889/files#r800206754 is still not resolved.

AspectInteractive2 commented:
> https://github.com/OpenRA/OpenRA/pull/19889/files#r800206754 is still not resolved.

Ok, is there anything that I can do to resolve this? From what I understand if this were an issue (which I highly doubt, given that we already use many functions from the same Math library where Math.Sin and Math.Cos reside, and have tested this on all 3 OSes), it would only be an issue for older Macs that use compat mode. So short of buying an old Mac, how can this be tested? Through a VM? More guidance here is needed.

abcdefg30 commented:
If I understood @RoosterDragon correctly the idea is to replace the two sinus and cosinus calls by a single `SinCos` call.

AspectInteractive2 commented:
> If I understood @RoosterDragon correctly the idea is to replace the two sinus and cosinus calls by a single `SinCos` call.

Oh I see what's happened. He has mentioned SinCos, probably referring to [this SinCos function, which is only .NET 6 compatible (no Xamarin is mentioned)](https://docs.microsoft.com/en-us/dotnet/api/system.math.sincos?view=net-6.0#system-math-sincos(system-double)). Perhaps he glanced the code and suspected I used this function, in which case there would indeed be a problem. However I only used Math.Sin and Math.Cos, which are much older and have support for numerous platforms (as shown earlier), so I think this can be treated as resolved.

abcdefg30 commented:
The idea was to change from Math.Sin and Math.Cos to Math.SinCos, but if we can't do that due to Mono I guess it is resolved indeed.

AspectInteractive2 commented:
> The idea was to change from Math.Sin and Math.Cos to Math.SinCos, but if we can't do that due to Mono I guess it is resolved indeed.

Oh, right, because it is slightly faster. Yes, well if it gets implemented in Mono at a later date this can be an quick PR, since it is easy to find the method and replace it once the compatibility exists.

penev92 commented:
I have a few small things I want to clear up:
 - You add a bunch of unnecessary line breaks to method signatures, like the entire `SpriteRenderer.cs` and `TerrainSpriteLayer.cs`.
 - You add a bunch of float default values with `0F` while we consistently use the lowercase `0f` everywhere else.
 - The PR description is now outdated and misleading after it was reworked.

And then a bigger one - this doesn't fix the jerking carryall on D2k's shellmap. It improves it, but doesn't entirely fix it. Since there are 1024 interpolated facings for it now I doubt the issue is a lack of facings. Probably an issue with the facing calculation?

Dzierzan commented:
Cool, it's exactly what I needed for OpenSA mod to make it even more complete.

AspectInteractive2 commented:
> I have a few small things I want to clear up:
> 
> * You add a bunch of unnecessary line breaks to method signatures, like the entire `SpriteRenderer.cs` and `TerrainSpriteLayer.cs`.

I suggest you have a read through all of the feedback above, as this has been both addressed and resolved multiple times already. All line breaks are now 120 characters, which is the industry standard. I know you prefer longer line breaks, though that is not convention and means scrolling is required on smaller and laptop screens in particular.


> * You add a bunch of float default values with `0F` while we consistently use the lowercase `0f` everywhere else.

Ok, normally I do prefer using the capital to differentiate from variable names which tend to be lowercase, but I can understand wanting to keep the code consistent, so I have amended these to use "0f" instead of "0F".

> * The PR description is now outdated and misleading after it was reworked.

I will amend this.

> 
> And then a bigger one - this doesn't fix the jerking carryall on D2k's shellmap. It improves it, but doesn't entirely fix it. Since there are 1024 interpolated facings for it now I doubt the issue is a lack of facings. Probably an issue with the facing calculation?

Yes it is not entirely possible to fix the jerking through 2D rotation alone, because the sprites are taken from a 3D model with a Z-axis, a 2D rotation won't be able to perfectly mimic 3D rotation. The only way to fully alleviate this currently is to use a 3D model of the carry-all and export 1024 facings. Eventually we may get to a stage where 3D models can be imported and rotated in 3D (i.e. including the z-axis), but this is a long way away. There isn't any issue with the facing calculation, as it works perfectly on sprites that are almost 2D (migs for example), and only shows slight jerkiness with more 3D-style sprites, most likely due to the reason above.

pchote commented:
OpenRA conventions are that general code style changes shouldn't be mixed in the same commit as non-trival logic changes (unless it is the same code that is being changed). This makes it harder to review and investigate any regressions when looking back (via git blame etc).

pchote commented on OpenRA.Test/OpenRA.Game/GraphicUtilTest.cs:
```diff
@@ -0,0 +1,67 @@
+		public void RotatePointTest()
+		{
+			var angleToRotate = new WAngle(128);
+			var oneSidedThreshold = new float3(0.5F, 0.5F, 0.5F);
```
`0.5f`

AspectInteractive2 replied:
Thanks, fixed.

pchote commented on OpenRA.Test/OpenRA.Game/GraphicUtilTest.cs:
```diff
@@ -0,0 +1,67 @@
+							rotatedPoint.Y <= expectedPoint.Y + oneSidedThreshold.Y &&
+							rotatedPoint.Y >= expectedPoint.Y - oneSidedThreshold.Y);
+
+				System.Console.WriteLine($"rotated: {rotatedPoint}, expected: {expectedPoint}, diff: {expectedPoint - rotatedPoint}");
```
Do we really want to be printing this?

AspectInteractive2 replied:
I found it useful during testing, but I can understand not wanting it there permanently. It is easy enough to add back if needed, so I have removed it.

AspectInteractive2 commented:
> OpenRA conventions are that general code style changes shouldn't be mixed in the same commit as non-trival logic changes (unless it is the same code that is being changed). This makes it harder to review and investigate any regressions when looking back (via git blame etc).

I did not make any style changes. The lines were getting very long once I added the rotation variable, necessitating a line break. The only line breaks that were added were specifically where the rotation variable was added and the line became too long. No other line breaks have been added or removed.

EDIT: I found two line breaks that I could revert where the rotation variable was not used (I believe I added the line break for consistency, since the same method was overloaded with line breaks earlier in the code), one in TerrainSpriteLayer.cs and another in SpriteRenderer.cs. Therefore the above now holds true, that there are no purely stylistic changes in this PR.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -20,12 +20,27 @@ public static class Util
+				var centerPoint = o + (size / 2);
+				var angleSin = (float)Math.Sin(-rotation);
+				var angleCos = (float)Math.Cos(-rotation);
+				a = RotatePoint(o, centerPoint, angleSin, angleCos);
```
I suspect that this won't be handling the z coordinate correctly.

I would expect the following to behave better, as well as being clearer and more efficient:
```csharp
// Rotate sprite if rotation angle is not equal to 0
if (rotation != 0f)
{
	var center = o + 0.5f * size;
	var angleSin = (float)Math.Sin(-rotation);
	var angleCos = (float)Math.Cos(-rotation);

	// Rotated offset for +/- x with +/- y
	var ra = 0.5f * new float3(
		size.X * angleCos - size.Y * angleSin,
		size.X * angleSin + size.Y * angleCos,
		(size.X * angleSin + size.Y * angleCos) * size.Z / size.Y);

	// Rotated offset for +/- x with -/+ y
	var rb = 0.5f * new float3(
		size.X * angleCos + size.Y * angleSin,
		size.X * angleSin - size.Y * angleCos,
		(size.X * angleSin - size.Y * angleCos) * size.Z / size.Y);

	a = center - ra;
	b = center + rb;
	c = center + ra;
	d = center - rb;
}
else
{
	a = o;
	b = new float3(o.X + size.X, o.Y, o.Z);
	c = new float3(o.X + size.X, o.Y + size.Y, o.Z + size.Z);
	d = new float3(o.X, o.Y + size.Y, o.Z + size.Z);
}
```

AspectInteractive2 replied:
The Z-coordinate is unaffected by the original code, since the rotation is only on the 2D plane (Yaw). I am not sure how your code above handles the Z-coord. With some brief testing it seems to work the same way as the original code. I can use your code above, however note that I haven't tested it as extensively, and wouldn't be able to reason about its validity (am relying on you for this). Also your code makes it impossible to re-use the RotatePoint() method on other points in other spaces (in particular if they don't comprise a 4 point quad), since it behaves differently based on which coordinate is being rotated. If you are okay with these caveats then I can proceed.

pchote replied:
The point is that the z coordinate *needs to be* affected, since the sprite can be projected onto an inclined plane in z.

pchote replied:
The issue can be demonstrated by setting `this.rotation = rotation + new WAngle(128);` in the `SpriteRenderable` ctor and then looking at the refinery/weapons factory bibs in TS. You can see visually (and even more clearly using the depth preview) how they are clipped against the terrain in the current PR, but drawn correctly in my suggested version.

So yes please do proceed, and remove the `RotatePoint` method.

AspectInteractive2 replied:
I have resolved this by using your approach above. I moved it into another function called RotateQuad(), so it can both be reused and also tested. I also replaced the previous test function with a RotateQuadTest() function, though admittedly I did not test the Z coordinate here as I am not entirely sure how that is expected to work.

pchote reviewed (changes requested):
`SpriteRenderable.RenderDebugGeometry` and `SpriteRenderable.ScreenBounds` will also need to be updated to account for rotation.

abcdefg30 commented on OpenRA.Game/Graphics/Animation.cs:
```diff
@@ -50,18 +50,22 @@ public Animation(World world, string name, Func<WAngle> facingFunc, Func<bool> p
 			var alpha = CurrentSequence.GetAlpha(CurrentFrame);
-			var imageRenderable = new SpriteRenderable(Image, pos, offset, CurrentSequence.ZOffset + zOffset, palette, CurrentSequence.Scale, alpha, float3.Ones, tintModifiers, IsDecoration);
+			var (image, rotation) = CurrentSequence.GetSpriteWithRotation(CurrentFrame, facingFunc());
+			var imageRenderable = new SpriteRenderable(image, pos, offset, CurrentSequence.ZOffset + zOffset, palette,
+													   CurrentSequence.Scale, alpha, float3.Ones, tintModifiers, IsDecoration, rotation);
```
Since @penev92 is now the second person complaining about line breaks, can we change these back? We should at least fix the indentation...

pchote replied:
Thirded.

AspectInteractive2 replied:
Ok, I've put CurrentSequence.Scale on the same line, which brings the line to 132 characters. I don't think we should exceed this, as most sites I've seen recommend 120 character lines, anything more than 130 characters will guarantee scrolling is required on 16" laptop screens. I have to say there seems to be a disconnect between the stringency of these line length rules, and the lack of any policy that I can follow, e.g. keep lines to 120 or 130 characters. A lot of this back and forth would have been resolved if there was a clear rule on this. The same can be said for indentation.

On the point of indentation, I am not sure what you mean by "fix" the indentation. Github is not showing it correctly above, in the editor, the parameters fall nicely underneath each other. I don't think it would be a good idea to move the indentation to the left, as then it may look like a separate new line rather than a continuation of the previous line. Though open to ideas on this.

![image](https://user-images.githubusercontent.com/1495872/154843120-03398840-b79b-4f93-8123-a7fc3facb681.png)

reaperrr replied:
We have a stylecop rule to not exceed 180 characters I think, other than that we try to keep things on one line where possible.

We usually only indent line breaks by a single tab, which is normally enough to show that it's not a separate new line.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -248,6 +266,41 @@ public static float[] MatrixMultiply(float[] lhs, float[] rhs)
+				size.X * angleSin - size.Y * angleCos,
+				(size.X * angleSin - size.Y * angleCos) * size.Z / size.Y);
+
+			return new float3[]
```
This is a hot code path, so we should be avoiding unnecessary allocations.

AspectInteractive2 replied:
Ok, but if it is inline then it's not testable. Would assigning the values to a,b,c,d as `out` variables directly work instead?

```c#
public static float3[] RotateQuad(float3 center, float3 size, float rotation, out float3 a, out float3 b, out float3 c, out float3 d)
{
    ...
    a = center - ra;
    b = center + rb;
    ...
}
```

pchote replied:
This is really basic code, so I don't think we should be making it complicated just to satisfy a desire for testing.

AspectInteractive2 replied:
Ok, I've removed the method and test case, and inlined the code just as it is in your snippet above.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -248,6 +266,41 @@ public static float[] MatrixMultiply(float[] lhs, float[] rhs)
+		/// <param name="size"> A float3 containing the X, Y, and Z lengths of the quad</param>
+		/// <param name="rotation">The number of radians to rotate by</param>
+		/// <returns>An array of four vertices representing the rotated quad (top-left, top-right, bottom-right, bottom-left)</returns>
+		#pragma warning disable SA1025 // Code should not contain multiple whitespace in a row
```
This shouldn’t be necessary.

AspectInteractive2 replied:
Have removed along with the method that it belonged to.

pchote commented:
Still missing:

> `SpriteRenderable.RenderDebugGeometry` and `SpriteRenderable.ScreenBounds` will also need to be updated to account for rotation.

AspectInteractive2 commented:
> Still missing:
> 
> > `SpriteRenderable.RenderDebugGeometry` and `SpriteRenderable.ScreenBounds` will also need to be updated to account for rotation.

Yeah I know, will look at it tomorrow, doing some other things today.

AspectInteractive2 commented:
> Still missing:
> 
> > `SpriteRenderable.RenderDebugGeometry` and `SpriteRenderable.ScreenBounds` will also need to be updated to account for rotation.

I've updated both `RenderDebugGeometry()` and `ScreenBounds()` for both `SpriteRenderable` and `UISpriteRenderable` to account for the rotation. However I don't know for certain if the `ScreenBounds()` code works, as I don't know how to reliably test it, but logically it seems correct (and doesn't present any issues in-game).

reaperrr commented:
> Error: /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Graphics/Animation.cs(61,132): error SA1001: Commas should be followed by whitespace. [/home/runner/work/OpenRA/OpenRA/OpenRA.Game/OpenRA.Game.csproj]

AspectInteractive2 commented:
> > Error: /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Graphics/Animation.cs(61,132): error SA1001: Commas should be followed by whitespace. [/home/runner/work/OpenRA/OpenRA/OpenRA.Game/OpenRA.Game.csproj]

This has now been fixed. For some reason Visual Studio was not showing this error on build or showing any warnings. No idea why, but after restarting Visual Studio I seemed to be able to see the warning and was able to remove the trailing white space.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -94,21 +116,42 @@ public void Render(WorldRenderer wr)
-			return new Rectangle((int)screenOffset.X, (int)screenOffset.Y, (int)sprite.Size.X, (int)sprite.Size.Y);
+			if (rotation == WAngle.Zero)
+				return new Rectangle((int)screenOffset.X, (int)screenOffset.Y, (int)sprite.Size.X, (int)sprite.Size.Y);
+			else
```
Can skip the `else` here since both branches return, which saves an indent.

AspectInteractive2 replied:
Fixed

RoosterDragon commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -94,21 +116,42 @@ public void Render(WorldRenderer wr)
+				var maxX = rotatedQuad[0].X;
+				var minY = rotatedQuad[0].Y;
+				var maxY = rotatedQuad[0].Y;
+				for (var i = 0; i < rotatedQuad.Length; i++)
```
`var i = 1` since we already dealt with 0?

AspectInteractive2 replied:
Fixed

RoosterDragon commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -94,21 +116,42 @@ public void Render(WorldRenderer wr)
+				var maxY = rotatedQuad[0].Y;
+				for (var i = 0; i < rotatedQuad.Length; i++)
+				{
+					minX = rotatedQuad[i].X < minX ? rotatedQuad[i].X : minX;
```
`minX = Math.Min(minX, rotatedQuad[i].X);`, etc?

AspectInteractive2 replied:
At least in JS, If statement is more performant than Math.Min/Max. I suspect it is similar in C#. And given this is rendering code I think we should use the most performant function possible.
https://stackoverflow.com/questions/1232345/speed-and-style-of-math-max-vs-ternary-operator-in-javascript

pchote replied:
C# isn't JS!

I would hope that the JIT would be inlining `Math.Max` making the performance of the two options equivalent - so prefer readability unless you have profiling that shows otherwise!

AspectInteractive2 replied:
After researching a bit more it seems in C# it is either comparable or slightly better with `Math.Max`, so I will make the switch
https://stackoverflow.com/questions/5478877/math-max-vs-inline-if-what-are-the-differences

AspectInteractive2 replied:
This has been updated now.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -94,21 +116,42 @@ public void Render(WorldRenderer wr)
+					maxY = rotatedQuad[i].Y > maxY ? rotatedQuad[i].Y : maxY;
+				}
+
+				return new Rectangle((int)minX, (int)minY, (int)(maxX - minX), (int)(maxY - minY));
```
I think this logic (and to be fair the existing logic for rotation of 0) might be losing pixels when dealing with fractions.

Consider this 1-dimensional example. A range from 0.9 to 1.1 has a bounding range of 0 to 2. However the current code would give 0 to 0 as the result!

I think this is needed to correct things:
```c#
new Rectangle(
    (int)minX,
    (int)minY,
    (int)Math.Ceiling(maxX - (int)minX),
    (int)Math.Ceiling(maxY - (int)minY));
```

RoosterDragon replied:
On a similar theme `Rectangle` has an *exclusive* upper bound so we might need a `+1` on the width and height as well - although that would also apply to the existing code.

AspectInteractive2 replied:
Good point about needing Math.Ceiling, I have since made the change. Re: the rectangle having an exclusive upper bound however, I had a look and it doesn't seem to be the case, e.g. `Rectangle.Right` is `X + Width`, `Rectangle.Bottom` is `Y + Height`, which would produce the correct result even with a Width of 1 or Height of 1. Unless other parts of the rendering code are doing `Rectangle.Right - 1`, or `Rectangle.Bottom - 1` (admittedly I have not looked as I am not too familiar with how ScreenBounds() is used. Also I omitted the `(int)` on `minX` and `minY` since this would lead to a further +1 added to the width and height, e.g. `Math.Ceiling(1.1 - 0.9)` would become `Math.Ceiling(1.1 - 0)`, making an object with a width of 1 return 2 for example. If we need to pad these values I think it's better to simply add +1, +2, or whatever the padding is required, but based on how this was coded previously by passing `size` with no padding, I don't think it is necessary.

RoosterDragon replied:
I would need to verify the Rectangle bounds, so let's ignore the +1 faff on that for now. The `(int)` cast is needed though. The 0.9-1.1 example has a width of only 0.2, but the bounding area is still 0 to 2. A result of 0 to 1 means we're slicing off some of the area!

AspectInteractive2 replied:
Ah good point, I see the issue now. Even though a Width of `1` should normally cover a range between 0.9 to 1.1, the `minX` is actually 0 because `minX` is being cast from 0.9 to 0 via `(int)minX`. Therefore from a starting point of 0 a Width of `2` is needed to cover the point 1.1. I have refactored the above slightly to make it easier to understand, by keeping the `(int)minX` outside of the `Ceiling()` function, to make it easier to draw the parallel that I initially missed between this `(int)minX` and the first parameter `(int)minX` passed to this Rectangle function, since this is the precise reason we need `(int)minX` in the second case.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteRenderable.cs:
```diff
@@ -94,21 +116,42 @@ public void Render(WorldRenderer wr)
 		{
 			var screenOffset = ScreenPosition(wr) + sprite.Offset;
-			return new Rectangle((int)screenOffset.X, (int)screenOffset.Y, (int)sprite.Size.X, (int)sprite.Size.Y);
+			if (rotation == WAngle.Zero)
```
Feels like extracting a method `Util.BoundingRectangle(offset, size, rotation)` or similar might be sensible to remove duplication?

AspectInteractive2 replied:
Good idea, this has now been done.

AspectInteractive2 commented:
One minor change not mentioned above in the latest commit. I have now updated `UISpriteRenderable.cs` to be consistent with `SpriteRenderable.cs`, as previously it was using `float` for the `rotation` parameter, where as `SpriteRenderable.cs` is now using `WAngle`. They are now both using `WAngle`.

RoosterDragon commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -191,6 +223,69 @@ public static void FastCopyIntoSprite(Sprite dest, Png src)
+				(int)minX,
+				(int)minY,
+				(int)Math.Ceiling(maxX) - (int)minX,
+				(int)Math.Ceiling(minY) - (int)minY);
```
Typo here, uses `minY` twice.

AspectInteractive2 replied:
Thanks, fixed.

AspectInteractive2 commented:
> One minor change not mentioned above in the latest commit. I have now updated `UISpriteRenderable.cs` to be consistent with `SpriteRenderable.cs`, as previously it was using `float` for the `rotation` parameter, where as `SpriteRenderable.cs` is now using `WAngle`. They are now both using `WAngle`.

I've now reverted this change due to some feedback that `UISpriteRenderable.cs` shouldn't be using world logic classes such as `WAngle`. However in order to enable `Util.BoundingRectangle()` to be used by both classes, this function now uses a `float` instead of a `WAngle`, which is actually the right approach since the `Util` class does not seem to be using any world logic classes at present. All this required is a cast from `SpriteRenderable.cs`'s WAngle to float using the `RendererRadians()` function, and since this was already being done within `Util.BoundingRectangle()`, there is no added performance loss here.

pchote commented on OpenRA.Game/Graphics/UISpriteRenderable.cs:
```diff
@@ -40,6 +42,14 @@ public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos,
+		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette, float scale)
+			: this(sprite, effectiveWorldPos, screenPos, zOffset, palette, scale, 1f, 0f) { }
+
+		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette, float scale, float alpha)
```
This ctor is not used?

AspectInteractive2 replied:
Yes however you will note this is the original ctor before rotation was implemented, as it has both scale and alpha parameters, while setting rotation to 0f. Note also that float and alpha are no longer the last two parameters, so a default rotation of 0f now needs to be passed when using the original ctor. I have reworked the calls to UISpriteRenderable to take the rotation parameter, which is why this ctor is not being referenced, however I think it is worth keeping this for any cases where rotation is not needed, since it is the only way to call UISpriteRenderable without passing in a rotation.

pchote commented on OpenRA.Game/Graphics/UISpriteRenderable.cs:
```diff
@@ -22,8 +22,9 @@ public class UISpriteRenderable : IRenderable, IPalettedRenderable, IFinalizedRe
+		readonly float rotation = 0f;
 
-		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette, float scale = 1f, float alpha = 1f)
+		public UISpriteRenderable(Sprite sprite, WPos effectiveWorldPos, int2 screenPos, int zOffset, PaletteReference palette, float scale, float alpha, float rotation)
```
Why replace the default values with additional ctors?

AspectInteractive2 replied:
Because it is otherwise not possible to pass in rotation, since parameters with default values must always be last (scale and alpha above having default value of 1f).

pchote replied:
?

AspectInteractive2 replied:
https://www.geeksforgeeks.org/c-sharp-optional-parameters/

>The optional parameters are always defined at the end of the parameter list. Or in other words, the last parameter of the method, constructor, etc. is the optional parameter.

AspectInteractive2 replied:
Ok this should actually work with the default parameter `rotation = 0f` at the end. Have updated the code.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -191,6 +223,69 @@ public static void FastCopyIntoSprite(Sprite dest, Png src)
+		/// <param name="size">A float3 containing the X, Y, and Z lengths of the quad</param>
+		/// <param name="rotation">The number of radians to rotate by</param>
+		/// <returns>An array of four vertices representing the rotated quad (top-left, top-right, bottom-right, bottom-left)</returns>
+		public static float3[] RotateQuad(float3 tl, float3 size, float rotation)
```
IMO its unintuitive for this method to take the top-left corner and size, but then rotate about its center (and in fact I had to delete earlier review comments where I pointed out what I thought were bugs before I read this). It would be more natural to take either the center and its size, or the top-left and bottom-right corners.

AspectInteractive2 replied:
I see what you mean, however using the top-left corner fits nicely with the `SpriteRenderable` call, where top-left corner is easily available. Also while `SpriteRenderable` does not pass in Size directly (it passes in bottom-right minus top-left), `UISpriteRenderable` and the `BoundingRectangle()` function both pass in `sprite.Size` as the second parameter, so changing this to the bottom-right corner would add further calculations. So while it might be a little unintuitive, I think the alternatives would require more calculations than at present.

pchote replied:
That doesn’t make sense… the first this this method does is to calculate the center, since it can’t work with the top-left directly.

AspectInteractive2 replied:
> That doesn’t make sense… the first this this method does is to calculate the center, since it can’t work with the top-left directly.

Not more calculations performance wise, but it would mean the center calculation would exist in more places than it currently does, since each function call would need to calculate it beforehand. I think this is messier and more prone to error, but if you really want it I guess it can be done. Though given this doesn't change the functionality at all I wonder whether this might be overcomplicating what is a fairly simple function.

pchote replied:
Ok. Can we please instead change the summary to something like
```
/// <summary>Rotates a quad about its center in the x-y plane.</summary>
```
?

AspectInteractive2 replied:
Sure, thanks for this. Done.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -191,6 +223,69 @@ public static void FastCopyIntoSprite(Sprite dest, Png src)
+		/// <param name="offset">The top left vertex of the object</param>
+		/// <param name="size">A float 3 containing the X, Y, and Z lengths of the object</param>
+		/// <param name="rotation">The angle to rotate the object by (use WAngle.Zero if there is no rotation)</param>
+		public static Rectangle BoundingRectangle(float3 offset, float3 size, float rotation)
```
Likewise here.

pchote commented on OpenRA.Game/Graphics/Util.cs:
```diff
@@ -191,6 +223,69 @@ public static void FastCopyIntoSprite(Sprite dest, Png src)
+		/// </summary>
+		/// <param name="offset">The top left vertex of the object</param>
+		/// <param name="size">A float 3 containing the X, Y, and Z lengths of the object</param>
+		/// <param name="rotation">The angle to rotate the object by (use WAngle.Zero if there is no rotation)</param>
```
`WAngle.Zero` is not a valid float!

AspectInteractive2 replied:
Ah, this must have been a classic case of documentation not being maintained with the code. I believe this function previously was using `WAngle` rather than floats. Have updated it now.

pchote commented:
The large interpolated facings don't play nicely with carryalls in d2k:

Before (no interpolated facings):
<img src="https://user-images.githubusercontent.com/167819/166104566-5fb657b4-fece-4dd5-b781-fdd756c59024.png">
 
After:
<img src="https://user-images.githubusercontent.com/167819/166104572-9b9da72b-4734-4bbc-8b8f-803c2610308a.png">

It looks like the carryall is being slightly rotated while the cargo is remaining fixed in the cardinal direction.

Testcase: The brown carryall in the shellmap.

AspectInteractive2 commented:
> The large interpolated facings don't play nicely with carryalls in d2k:
> 
> Before (no interpolated facings):
> After: 
> It looks like the carryall is being slightly rotated while the cargo is remaining fixed in the cardinal direction.
> 
> Testcase: The brown carryall in the shellmap.

I suspect this is caused by the cargo not having the same InterpolatedFacings trait that the carry-all does, and it not being able to be inherited. Will investigate and report back.

AspectInteractive2 commented:
Ok so from what I can tell it seems that this is due to the Harvester still having only 32 facings, while the carry-all now has 1024. The fact that the carry-all is slightly tilted only in the interpolated facings version, shows that the tilt in the "After" image above is one of the interpolated facings, which the harvester does not have. Therefore adding more facings to the harvester would solve the issue, but I see that it has the "UseClassicFacings" trait, so it would not be possible unless it was changed to not use classic facings (classic facings are required to only have 32 facings).

anvilvapre commented:
#19582.

Mailaender committed:
Added rotation logic to the renderer to enable the use of Interpolated Facings.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/898ad33df92f998d2b28023643a2f9bb19b20a8d...220ca7f4740472c9f36bee7741d7e5c8765c5cf8)
